### PR TITLE
Solve issue 1865, add is_trail & is_acyclic path functions

### PR DESCRIPTION
**Pull request by andyfengHKU**
#1865

### CODE REVIEW

File: src/expression_evaluator/node_rel_evaluator.cpp [hunk 1]
```diff
@@ -9,6 +9,7 @@ namespace kuzu {
 namespace evaluator {
 
 void NodeRelExpressionEvaluator::evaluate() {
+    resultVector->resetAuxiliaryBuffer();
```
Comment (acquamarin) [inline src/expression_evaluator/node_rel_evaluator.cpp:12]: Guess we should move resultVector->resetAuxiliaryBuffer to sourceOperator::getNextTuple. 
So each sourceOperator should reset auxiliaryBuffer beforing writing to the vector.

File: src/include/function/path/path_function_executor.h [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+// Note: this executor is only used for isTrail and isAcyclic. So we add some ad-hoc optimization
+// into executor, e.g. internalIDSet. A more general implementation can be done once needed. But
+// pay attention to the performance drop. Depends on how bad it becomes, we may want to implement
+// customized executors.
+struct UnaryPathExecutor {
+    static void executeNodeIDs(common::ValueVector& input, common::ValueVector& result) {
+        assert(
+            0 == common::StructType::getFieldIdx(&input.dataType, common::InternalKeyword::NODES));
+        auto nodesVector = common::StructVector::getFieldVector(&input, 0).get();
+        execute(*input.state->selVector, *nodesVector, 0, result);
```
Comment (acquamarin) [inline src/include/function/path/path_function_executor.h:33]: Add a comment for 0.

File: src/include/function/hash/hash_functions.h [hunk 3]
```diff
@@ -117,5 +117,13 @@ inline void Hash::operation(const std::unordered_set<std::string>& key, common::
     }
 }
 
+struct InternalIDHasher {
```
Comment (acquamarin) [inline src/include/function/hash/hash_functions.h:120]: I think this class in only used by unordered_set when computing hash for internalID and it is only used in path_function_executor. Maybe move to the path_function_executor.h?

**Review by acquamarin — APPROVED** [review]

File: src/expression_evaluator/node_rel_evaluator.cpp (hunk 1 above)
Comment (andyfengHKU) [reply src/expression_evaluator/node_rel_evaluator.cpp:12]: As we discussed, this is not as trivial as reset on source operator. All expression evaluator should reset.

File: src/include/function/hash/hash_functions.h (hunk 3 above)
Comment (andyfengHKU) [reply src/include/function/hash/hash_functions.h:120]: Also used for recursive join. So refactored into common

**Pull request merged**